diff -Naur fbset-2.1.orig/fbset.c fbset-2.1/fbset.c
--- fbset-2.1.orig/fbset.c	2003-11-18 12:45:24.000000000 +0100
+++ fbset-2.1/fbset.c	2003-11-18 12:45:38.000000000 +0100
@@ -833,7 +833,8 @@
 static void Usage(void)
 {
     puts(VERSION);
-    Die("\nUsage: %s [options] [mode]\n\n"
+    printf(
+	"\nUsage: %s [options] [mode]\n\n"
 	"Valid options:\n"
 	"  General options:\n"
 	"    -h, --help         : display this usage information\n"
@@ -888,6 +889,7 @@
 	"    -step <value>      : step increment (in pixels or pixel lines)\n"
 	"                         (default is 8 horizontal, 2 vertical)\n",
 	ProgramName);
+	exit(0);
 }
 
 
